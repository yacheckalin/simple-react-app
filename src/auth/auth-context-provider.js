import React, { useState, useContext } from "react";
import { auth } from "./helpers";

const context = {
  isAuthorised: false,
};

export const AuthContext = React.createContext(context);
export const useAuthContext = () => useContext(AuthContext);

const errorResponse = JSON.stringify({
  result: "error",
  error:
    "\u041d\u0435\u0432\u0435\u0440\u043d\u044b\u0435 \u043f\u0430\u0440\u0430\u043c\u0435\u0442\u0440\u044b",
});
const successResponse = JSON.stringify({ result: "ok" });

const AuthContextProvider = ({ children }) => {
  const [isAuthorised, setIsAuthorised] = useState(true);
  const [loginError, setLoginError] = useState("");

  const login = ({ email: login, password }) => {
    //TODO: change for axios request
    let resp = auth({ email: login, password })
      ? successResponse
      : errorResponse;
    const response = Promise.resolve(resp);

    response.then((response) => {
      const res = JSON.parse(response);
      if (res.error) {
        setLoginError(res.error);
        setIsAuthorised(false);
      } else {
        setIsAuthorised(true);
      }
    });
  };

  const logout = ({ email, password }) => {
    //TODO: change for axios request
    const response = Promise.resolve(successResponse);

    response.then((resp) => {
      const res = JSON.parse(resp);

      if (!res.error) {
        setLoginError("");
        setIsAuthorised(false);
      }
    });

    setIsAuthorised(false);
    setLoginError("");
  };

  const resetLoginError = () => {
    setLoginError("");
  };

  return (
    <AuthContext.Provider
      value={{
        isAuthorised,
        setIsAuthorised,
        login,
        logout,
        loginError,
        resetLoginError,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export default AuthContextProvider;
