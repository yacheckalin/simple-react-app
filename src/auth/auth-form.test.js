import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";

import AuthForm from "./auth-form";
import * as AuthContext from "./auth-context-provider";

describe("<AuthForm />", () => {
  test("renders and displays properly", () => {
    const wrapper = shallow(<AuthForm />);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test("form elements renders properly", () => {
    const wrapper = shallow(<AuthForm />);

    expect(wrapper.find("input")).toHaveLength(2);
    expect(wrapper.find("button")).toHaveLength(1);
  });

  test("validate email shows error message", () => {
    const contextValues = { isAuthorised: false, resetLoginError: jest.fn() };
    jest
      .spyOn(AuthContext, "useAuthContext")
      .mockImplementation(() => contextValues);
    const wrapper = shallow(<AuthForm />);

    wrapper.find("#email").simulate("change", { target: { value: "test" } });
    expect(wrapper.find("#email").hasClass("error")).toBeTruthy();

    expect(wrapper.find("#email-error").text().trim()).toBe("Неверный email");
  });

  test("validate email", () => {
    const contextValues = {
      isAuthorised: false,
      loginError: "",
      resetLoginError: jest.fn(),
    };
    jest
      .spyOn(AuthContext, "useAuthContext")
      .mockImplementation(() => contextValues);

    const wrapper = shallow(<AuthForm />);
    wrapper
      .find("#email")
      .simulate("change", { target: { value: "test@test.ss" } });
    expect(wrapper.find("#email").hasClass("error")).toBeFalsy();

    expect(wrapper.find("#email-error").text().trim()).toBe("");
  });

  test("validate password shows error message", () => {
    const contextValues = { isAuthorised: false, resetLoginError: jest.fn() };
    jest
      .spyOn(AuthContext, "useAuthContext")
      .mockImplementation(() => contextValues);
    const wrapper = shallow(<AuthForm />);

    wrapper
      .find("#password")
      .simulate("change", { target: { value: "test2" } });
    expect(wrapper.find("#password").hasClass("error")).toBeTruthy();

    expect(wrapper.find("#password-error").text().trim()).toBe(
      "Неверный пароль"
    );
  });

  test("validate password", () => {
    const contextValues = {
      isAuthorised: false,
      loginError: "",
      resetLoginError: jest.fn(),
    };
    jest
      .spyOn(AuthContext, "useAuthContext")
      .mockImplementation(() => contextValues);

    const wrapper = shallow(<AuthForm />);
    wrapper
      .find("#password")
      .simulate("change", { target: { value: "PasswordSecured_" } });
    expect(wrapper.find("#password").hasClass("error")).toBeFalsy();

    expect(wrapper.find("#password-error").text().trim()).toBe("");
  });

  test("complete validation fail", () => {
    const contextValues = {
      isAuthorised: false,
      loginError: "Form Validation Error",
      resetLoginError: jest.fn(),
      login: jest.fn(),
    };
    jest
      .spyOn(AuthContext, "useAuthContext")
      .mockImplementation(() => contextValues);

    const wrapper = shallow(<AuthForm />);
    wrapper
      .find("#email")
      .simulate("change", { target: { value: "test@email.com" } });
    wrapper
      .find("#password")
      .simulate("change", { target: { value: "Passw_" } });
    wrapper.find("button").simulate("click", { preventDefault: () => {} });
    expect(wrapper.find("#form-error").text()).toBe("Form Validation Error");
  });
  test("complete validation successfully complete", () => {
    const contextValues = {
      isAuthorised: false,
      loginError: "",
      resetLoginError: jest.fn(),
      login: jest.fn(),
    };
    jest
      .spyOn(AuthContext, "useAuthContext")
      .mockImplementation(() => contextValues);

    const wrapper = shallow(<AuthForm />);
    wrapper
      .find("#email")
      .simulate("change", { target: { value: "test@email.com" } });
    wrapper
      .find("#password")
      .simulate("change", { target: { value: "PasswordSecured" } });
    wrapper.find("button").simulate("click", { preventDefault: () => {} });
    expect(wrapper.find("#form-error").text()).toBe("");
  });
});
