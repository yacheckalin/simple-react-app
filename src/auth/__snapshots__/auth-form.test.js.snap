// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthForm /> renders and displays properly 1`] = `
<styled.div>
  <styled.div>
    <span>
      login.header
    </span>
  </styled.div>
  <styled.div>
    <styled.div>
      <styled.div>
        <p>
           
          login.login
           
        </p>
        <input
          className="form-input false"
          id="email"
          onChange={[Function]}
          type="text"
          value=""
        />
        <p
          id="email-error"
        >
           
        </p>
      </styled.div>
    </styled.div>
    <styled.div>
      <styled.div>
        <p>
           
          login.password
           
        </p>
        <input
          className="form-input false"
          id="password"
          onChange={[Function]}
          type="password"
          value=""
        />
        <p
          id="password-error"
        >
           
        </p>
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div>
    <button
      onClick={[Function]}
    >
      <span>
        login.enter
      </span>
      <span>
        <IoIosArrowRoundForward
          color="white"
          size="2em"
        />
      </span>
    </button>
    <p
      id="form-error"
    >
      login.formError
    </p>
  </styled.div>
</styled.div>
`;
